### PR TITLE
account registration

### PR DESCRIPTION
**Pull request by ChFlick**
## Summary of change

Add user registration

## Additional context

<summary>Login screen: Added a link to the registration</summary>
<details>

image
</details>

<summary>Registration screen:</summary>
<details>

image
</details>

## Related issue

#140

### CODE REVIEW

**Comment by knjk04** [comment]
Thanks for this! Could you add an image of the UI change(s) to the additional context section? Sometimes, it looks different on my device, so it would be good to say what it looks like on yours.

**Comment by ChFlick** [comment]
Sure.

**Comment by knjk04** [comment]
Thanks, I'll review this shortly. The registration form's appearance looks really nice!

File: src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.AuthenticationServiceException;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.validation.ConstraintViolationException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:17]: As we are not using wildcard imports, could you change the occurrences of these static imports to use the qualified class name, instead?

**Comment by knjk04** [comment]
I've also added five commits that are mostly style changes

File: src/main/java/com/karankumar/bookproject/backend/entity/account/User.java [hunk 2]
```diff
@@ -54,10 +56,12 @@
     @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "uidgenerator")
     @SequenceGenerator(name = "uidgenerator", sequenceName = "uid_sequence")
     @Getter
+    @Setter(AccessLevel.NONE)
     private Long id;
 
     @NotNull
     @NotEmpty
+    @Length(min = 5, max = 64)
```
[the remark this answers is not in the document]
Comment (ChFlick) [reply src/main/java/com/karankumar/bookproject/backend/entity/account/User.java:64]: Nope, that was basically random, I'm not even sure if a lower bound is neccessary.

File: src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java (hunk 1 above)
Comment (ChFlick) [reply src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:17]: Do you consider the static imports wildcard imports?
As far as I know wildcard imports are `import a.b.c.*`.

In my opinion, static imports greatly improve the readability in tests, though that might be personal preference.

Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:17]: Personally, I consider any use of the wildcard (`*`) to be a wildcard import, but we did explicitly qualify this in the style guide (which Google also did with their Java style guide).

I can see where you're coming from. I would be happy to debate this over Gitter, as I'm always happy for us to change any of the style guide rules.

Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/service/UserServiceTest.java:17]: OK, my bad. We're agreeing on wildcard imports are, but I made a mistake by commenting on a static, non-wildcard import. Ignore me :)

**Review by knjk04 — APPROVED** [review]
Thanks a lot for this! This was really helpful and very well written, as usual!

**Pull request merged**